graphite_api/render/datalib.py

```diff
@@ -78,84 +78,108 @@ def __repr__(self):
             self.name, self.start, self.end, self.step)
 
 
-# Data retrieval API
-def fetchData(requestContext, pathExpr):
+class DataStore(object):
+    """
+    Simple object to store results of multi fetches.
+    Also aids in looking up data by pathExpressions.
+    """
+    def __init__(self):
+        self.paths = defaultdict(list)
+        self.data = defaultdict(list)
+
+    def add_path(self, path_expr, path):
+        """
+        Adds the path to the pathExpression.
+        """
+        # Only add if we dont already have the path
+        if path not in self.paths[path_expr]:
+            self.paths[path_expr].append(path)
```

Comment:
You could make `self.paths` a `defaultdict(set)` and just do `self.paths[path_expr].add(path)` without checking for membership. Paths aren't ordered anymore if we use a `set` but `get_paths` can be changed to return `sorted(self.paths[path_expr])` to keep rendering consistent across requests.